### PR TITLE
OverconstrainedError extends DOMException + algo fixes

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes https://github.com/w3c/mediacapture-main/issues/568.

I made the `constraint` argument always present and `""` in cases where the constraint name is either filtered out or undetermined. This matches Safari and Firefox at least (I wasn't able to provoke the absence of a constraint in Chrome).

I also fixed some bugs in the algorithms around `failedConstraint` in the process.

***
<a href="https://pr-preview.s3.amazonaws.com/jan-ivar/mediacapture-main/pull/746.html" title="Last updated on Nov 9, 2020, 11:23 PM UTC (61257f4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-main/746/87cb2f7...jan-ivar:61257f4.html" title="Last updated on Nov 9, 2020, 11:23 PM UTC (61257f4)">Diff</a>

### CODE REVIEW

**Comment by jan-ivar** [comment]
@henbos Regarding question in Thursday's meeting, `instanceof OverconstrainedError` should still work.

**Comment by saschanaz** [comment]
It seems `MediaStreamError` is now essentially just `DOMException`. Can it be removed here?

**Comment by jan-ivar** [comment]
> It seems `MediaStreamError` is now essentially just `DOMException`. Can it be removed here?

@saschanaz Can we return a base class interface in a WebIDL JS callback and have the following still work?
```js
navigator.mediaDevices.getUserMedia({video: {width: {min: 9999}}}, () => {}, e => {
  console.log(e.constraint); // width
});  
```
I've been working too much with dictionaries where this would be a problem.

**Comment by saschanaz** [comment]
Yes, it should work. A working example is [`EventHandler`](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler), where the argument type is the base class interface `Event`.

File: getusermedia.html [hunk 1]
```diff
@@ -2332,230 +2332,68 @@ <h2>MediaStreams in Media Elements</h2>
...
-          </tr>
-          <tr>
-            <td>Object.prototype.toString</td>
-            <td>19.1.3.6</td>
-          </tr>
-        </tbody>
-      </table>
-    </section>
+    <p>Some operations throw or fire <code>OverconstrainedError</code>. This is
+    an extension of <code>DOMException</code> that carries additional
```
Comment (dontcallmedom) [inline getusermedia.html:2336]:
```suggestion
    an extension of {{DOMException}} that carries additional
```

File: getusermedia.html [hunk 2]
```diff
@@ -2332,230 +2332,68 @@ <h2>MediaStreams in Media Elements</h2>
...
+              <li>
+                <p>Let <var>message</var> be the constructor's second argument.
+                </p>
+              </li>
+              <li>
+                <p>Let <var>e</var> be a new
+                <code><a>OverconstrainedError</a></code> object.</p>
+              </li>
+              <li>
+                <p>Invoke the <code><a>DOMException</a></code> constructor of
```
Comment (dontcallmedom) [inline getusermedia.html:2369]:
```suggestion
                <p>Invoke the {{DOMException}} constructor of
```

File: getusermedia.html [hunk 3]
```diff
@@ -2332,230 +2332,68 @@ <h2>MediaStreams in Media Elements</h2>
...
+                <p>Let <var>constraint</var> be the constructor's first
+                argument.</p>
+              </li>
+              <li>
+                <p>Let <var>message</var> be the constructor's second argument.
+                </p>
+              </li>
+              <li>
+                <p>Let <var>e</var> be a new
+                <code><a>OverconstrainedError</a></code> object.</p>
```
Comment (dontcallmedom) [inline getusermedia.html:2366]:
```suggestion
                {{OverconstrainedError}} object.</p>
```

File: getusermedia.html [hunk 4]
```diff
@@ -2332,230 +2332,68 @@ <h2>MediaStreams in Media Elements</h2>
...
-        <section>
-          <h2>OverconstrainedError.prototype.name</h2>
-          <p>The initial value of the name property of the prototype for the
-          OverconstrainedError constructor is
-          <code>"<b>OverconstrainedError</b>"</code>.</p>
-        </section>
+        <h2>Attributes</h2>
+        <dl data-link-for="OverconstrainedError" data-dfn-for="OverconstrainedError" class="attributes">
+          <dt><dfn data-idl><code>constraint</code></dfn> of type <span
+          class="idlAttrType"><a>DOMString</a></span>, readonly</dt>
```
Comment (dontcallmedom) [inline getusermedia.html:2391]:
```suggestion
          class="idlAttrType">{{DOMString}}</span>, readonly</dt>
```
(this will fix the CI error)

**Review by henbos — APPROVED** [review]

**Pull request merged**